community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,14 +67,64 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  private def iterateRelationships(row: ExecutionContext, state: QueryState, fromNode: Node, toNode: Node) = {
+
+    //check degree and iterate from the node with smaller degree
+    val fromDegree = state.query.nodeGetDegree(fromNode.getId, dir)
+    val toDegree = state.query.nodeGetDegree(toNode.getId, dir.reverse())
+    val (start, end, relationships) = if (fromDegree < toDegree)
+      (fromNode, toNode, state.query.getRelationshipsForIds(fromNode, dir, types.types(state.query)))
+    else
+      (toNode, fromNode, state.query.getRelationshipsForIds(toNode, dir.reverse(), types.types(state.query)))
+
+    if (toNode == null)
+      Iterator.empty
+    else {
+      new PrefetchingIterator[ExecutionContext] {
+        val connectedRelationships = ArrayBuffer.empty[Relationship]
+
+        override def fetchNextOrNull(): ExecutionContext = {
+          while (relationships.hasNext) {
+            val rel: Relationship = relationships.next()
+            val other = rel.getOtherNode(start)
+            if (end == other) {
+              connectedRelationships.append(rel)
+              return row.newWith2(relName, rel, toName, toNode)
+            }
+          }
+          registerConnection(start, end, connectedRelationships)
+          null
+        }
+      }
+    }.asScala
+  }
+
+  private def hasNoConnection(node1: Node, node2: Node) = {
+    val (n1, n2) = if (node1.getId < node2.getId) (node1, node2) else (node2, node1)
+    nonConnected.get(n1).exists(_ == n2)
+  }
+
+  private def registerConnection(node1: Node, node2: Node, rels: Seq[Relationship]) = {
+    val (n1, n2) = if (node1.getId < node2.getId) (node1, node2) else (node2, node1)
+    if (rels.isEmpty) nonConnected.put(n1, n2)
+    else connected += (n1, n2) -> rels
+  }
+
+  private def fromCache(row: ExecutionContext, fromNode: Node, toNode: Node) = {
+    val elem = if (fromNode.getId < toNode.getId) (fromNode, toNode) else (toNode, fromNode)
+    connected.get(elem).map {
+      case rels => rels.map(row.newWith2(relName, _, toName, toNode)).toIterator
```

Comment:
not sure what the perf characteristics of `toIterator()` is. I vaguely remember one of these methods does a toList internally. It should be a lazy evaluation.
Do we have to create new rows for each rel?